### PR TITLE
Fix memory leak in ResourceFormatSaverBinary::save

### LINKED ISSUE #30658: Memory leak when importing first time files
**Godot version:**
3.2 df62bfdd4

**Issue description:**
```
2,616 (216 direct, 2,400 indirect) bytes in 3 blocks are definitely lost in loss record 137 of 152
   at 0x5A7B74F: malloc (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
   by 0x38AFDE1: Memory::alloc_static(unsigned long, bool) (memory.cpp:85)
   by 0x38AFD8A: operator new(unsigned long, char const*) (memory.cpp:42)
   by 0x1E8B11C: FileAccess* FileAccess::_create_builtin<FileAccessUnix>() (file_access.h:77)
   by 0x3883AFE: FileAccess::create(FileAccess::AccessType) (file_access.cpp:49)
   by 0x3883BF6: FileAccess::create_for_path(String const&) (file_access.cpp:76)
   by 0x3883D03: FileAccess::open(String const&, int, Error*) (file_access.cpp:108)
   by 0x3972117: ResourceFormatSaverBinaryInstance::save(String const&, Ref<Resource> const&, unsigned int) (resource_format_binary.cpp:1763)
   by 0x3973E04: ResourceFormatSaverBinary::save(String const&, Ref<Resource> const&, unsigned int) (resource_format_binary.cpp:1965)
   by 0x398A032: ResourceSaver::save(String const&, Ref<Resource> const&, unsigned int) (resource_saver.cpp:119)
   by 0x23DBC40: ResourceImporterScene::import(String const&, String const&, Map<StringName, Variant, Comparator<StringName>, DefaultAllocator> const&, List<String, DefaultAllocator>*, List<String, DefaultAllocator>*, Variant*) (resource_importer_scene.cpp:1487)
   by 0x2039EB6: EditorFileSystem::_reimport_file(String const&) (editor_file_system.cpp:1783)
   by 0x203C967: EditorFileSystem::reimport_files(Vector<String> const&) (editor_file_system.cpp:1981)
   by 0x202F53A: EditorFileSystem::_update_scan_actions() (editor_file_system.cpp:590)
   by 0x2033525: EditorFileSystem::_notification(int) (editor_file_system.cpp:1148)
   by 0x203FC2F: EditorFileSystem::_notificationv(int, bool) (in /home/rafal/Pulpit/mojgodot/bin/godot.x11.tools.64)
   by 0x3772685: Object::notification(int, bool) (object.cpp:943)
   by 0x29BBBC4: SceneTree::_notify_group_pause(StringName const&, int) (scene_tree.cpp:973)
   by 0x29B9799: SceneTree::idle(float) (scene_tree.cpp:519)
   by 0x13CA518: Main::iteration() (main.cpp:1934)
   by 0x13997F0: OS_X11::run() (os_x11.cpp:3177)
   by 0x1389619: main (godot_x11.cpp:56)
``` 

**Minimal reproduction project:**
[FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF.zip](https://github.com/godotengine/godot/files/3403492/FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF.zip)

### LINKED ISSUE #31120: Memory leak when importing wav and ogg files
**Godot version:**
3.2.dev.custom_build 7126654ea

**OS/device including version:**
Ubuntu 18.04

**Issue description:**
Part of leaks from below comes from this issue #30658
```
Direct leak of 72 byte(s) in 1 object(s) allocated from:
    #0 0x7f5d86a69b50 in __interceptor_malloc (/usr/lib/x86_64-linux-gnu/libasan.so.4+0xdeb50)
    #1 0xcad6b2a in Memory::alloc_static(unsigned long, bool) core/os/memory.cpp:85
    #2 0xcad6a26 in operator new(unsigned long, char const*) core/os/memory.cpp:42
    #3 0x44cb5fc in FileAccess* FileAccess::_create_builtin<FileAccessUnix>() core/os/file_access.h:77
    #4 0xca14dcd in FileAccess::create(FileAccess::AccessType) core/os/file_access.cpp:49
    #5 0xca15265 in FileAccess::create_for_path(String const&) core/os/file_access.cpp:77
    #6 0xca155ae in FileAccess::open(String const&, int, Error*) core/os/file_access.cpp:109
    #7 0xce7bf4c in ResourceFormatSaverBinaryInstance::save(String const&, Ref<Resource> const&, unsigned int) core/io/resource_format_binary.cpp:1763
    #8 0xce85571 in ResourceFormatSaverBinary::save(String const&, Ref<Resource> const&, unsigned int) core/io/resource_format_binary.cpp:1965
    #9 0xcedbfeb in ResourceSaver::save(String const&, Ref<Resource> const&, unsigned int) core/io/resource_saver.cpp:119
    #10 0x1bd9732 in ResourceImporterOGGVorbis::import(String const&, String const&, Map<StringName, Variant, Comparator<StringName>, DefaultAllocator> const&, List<String, DefaultAllocator>*, List<String, DefaultAllocator>*, Variant*) modules/stb_vorbis/resource_importer_ogg_vorbis.cpp:106
    #11 0x4e2eec5 in EditorFileSystem::_reimport_file(String const&) editor/editor_file_system.cpp:1781
    #12 0x4e391e4 in EditorFileSystem::reimport_files(Vector<String> const&) editor/editor_file_system.cpp:1979
    #13 0x4dfd388 in EditorFileSystem::_update_scan_actions() editor/editor_file_system.cpp:590
    #14 0x4e13a71 in EditorFileSystem::_notification(int) editor/editor_file_system.cpp:1163
    #15 0x4e46b81 in EditorFileSystem::_notificationv(int, bool) (/home/rafal/Pulpit/mojgodot/bin/godot.x11.tools.64s+0x4e46b81)
    #16 0xc5f70f2 in Object::notification(int, bool) core/object.cpp:943
    #17 0x8094048 in SceneTree::_notify_group_pause(StringName const&, int) scene/main/scene_tree.cpp:973
    #18 0x8085f01 in SceneTree::idle(float) scene/main/scene_tree.cpp:519
    #19 0x1501b07 in Main::iteration() main/main.cpp:1935
    #20 0x1421219 in OS_X11::run() platform/x11/os_x11.cpp:3178
    #21 0x13ac4e1 in main platform/x11/godot_x11.cpp:56
    #22 0x7f5d835ebb96 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x21b96)

Direct leak of 72 byte(s) in 1 object(s) allocated from:
    #0 0x7f5d86a69b50 in __interceptor_malloc (/usr/lib/x86_64-linux-gnu/libasan.so.4+0xdeb50)
    #1 0xcad6b2a in Memory::alloc_static(unsigned long, bool) core/os/memory.cpp:85
    #2 0xcad6a26 in operator new(unsigned long, char const*) core/os/memory.cpp:42
    #3 0x44cb5fc in FileAccess* FileAccess::_create_builtin<FileAccessUnix>() core/os/file_access.h:77
    #4 0xca14dcd in FileAccess::create(FileAccess::AccessType) core/os/file_access.cpp:49
    #5 0xca15265 in FileAccess::create_for_path(String const&) core/os/file_access.cpp:77
    #6 0xca155ae in FileAccess::open(String const&, int, Error*) core/os/file_access.cpp:109
    #7 0xce7bf4c in ResourceFormatSaverBinaryInstance::save(String const&, Ref<Resource> const&, unsigned int) core/io/resource_format_binary.cpp:1763
    #8 0xce85571 in ResourceFormatSaverBinary::save(String const&, Ref<Resource> const&, unsigned int) core/io/resource_format_binary.cpp:1965
    #9 0xcedbfeb in ResourceSaver::save(String const&, Ref<Resource> const&, unsigned int) core/io/resource_saver.cpp:119
    #10 0x60d98da in ResourceImporterWAV::import(String const&, String const&, Map<StringName, Variant, Comparator<StringName>, DefaultAllocator> const&, List<String, DefaultAllocator>*, List<String, DefaultAllocator>*, Variant*) editor/import/resource_importer_wav.cpp:548
    #11 0x4e2eec5 in EditorFileSystem::_reimport_file(String const&) editor/editor_file_system.cpp:1781
    #12 0x4e391e4 in EditorFileSystem::reimport_files(Vector<String> const&) editor/editor_file_system.cpp:1979
    #13 0x4dfd388 in EditorFileSystem::_update_scan_actions() editor/editor_file_system.cpp:590
    #14 0x4e13a71 in EditorFileSystem::_notification(int) editor/editor_file_system.cpp:1163
    #15 0x4e46b81 in EditorFileSystem::_notificationv(int, bool) (/home/rafal/Pulpit/mojgodot/bin/godot.x11.tools.64s+0x4e46b81)
    #16 0xc5f70f2 in Object::notification(int, bool) core/object.cpp:943
    #17 0x8094048 in SceneTree::_notify_group_pause(StringName const&, int) scene/main/scene_tree.cpp:973
    #18 0x8085f01 in SceneTree::idle(float) scene/main/scene_tree.cpp:519
    #19 0x1501b07 in Main::iteration() main/main.cpp:1935
    #20 0x1421219 in OS_X11::run() platform/x11/os_x11.cpp:3178
    #21 0x13ac4e1 in main platform/x11/godot_x11.cpp:56
    #22 0x7f5d835ebb96 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x21b96)

Indirect leak of 1040 byte(s) in 1 object(s) allocated from:
    #0 0x7f5d86a69f40 in realloc (/usr/lib/x86_64-linux-gnu/libasan.so.4+0xdef40)
    #1 0xcad75fa in Memory::realloc_static(void*, unsigned long, bool) core/os/memory.cpp:140
    #2 0x1ba9cf2 in CowData<wchar_t>::resize(int) core/cowdata.h:282
    #3 0x49f27ba in String::resize(int) core/ustring.h:158
    #4 0xc7d34f5 in String::operator+=(String const&) core/ustring.cpp:276
    #5 0xc7d31a4 in String::operator+(String const&) const core/ustring.cpp:252
    #6 0x4a24f0c in FileAccessUnix::_open(String const&, int) drivers/unix/file_access_unix.cpp:109
    #7 0xca1574a in FileAccess::open(String const&, int, Error*) core/os/file_access.cpp:110
    #8 0xce7bf4c in ResourceFormatSaverBinaryInstance::save(String const&, Ref<Resource> const&, unsigned int) core/io/resource_format_binary.cpp:1763
    #9 0xce85571 in ResourceFormatSaverBinary::save(String const&, Ref<Resource> const&, unsigned int) core/io/resource_format_binary.cpp:1965
    #10 0xcedbfeb in ResourceSaver::save(String const&, Ref<Resource> const&, unsigned int) core/io/resource_saver.cpp:119
    #11 0x60d98da in ResourceImporterWAV::import(String const&, String const&, Map<StringName, Variant, Comparator<StringName>, DefaultAllocator> const&, List<String, DefaultAllocator>*, List<String, DefaultAllocator>*, Variant*) editor/import/resource_importer_wav.cpp:548
    #12 0x4e2eec5 in EditorFileSystem::_reimport_file(String const&) editor/editor_file_system.cpp:1781
    #13 0x4e391e4 in EditorFileSystem::reimport_files(Vector<String> const&) editor/editor_file_system.cpp:1979
    #14 0x4dfd388 in EditorFileSystem::_update_scan_actions() editor/editor_file_system.cpp:590
    #15 0x4e13a71 in EditorFileSystem::_notification(int) editor/editor_file_system.cpp:1163
    #16 0x4e46b81 in EditorFileSystem::_notificationv(int, bool) (/home/rafal/Pulpit/mojgodot/bin/godot.x11.tools.64s+0x4e46b81)
    #17 0xc5f70f2 in Object::notification(int, bool) core/object.cpp:943
    #18 0x8094048 in SceneTree::_notify_group_pause(StringName const&, int) scene/main/scene_tree.cpp:973
    #19 0x8085f01 in SceneTree::idle(float) scene/main/scene_tree.cpp:519
    #20 0x1501b07 in Main::iteration() main/main.cpp:1935
    #21 0x1421219 in OS_X11::run() platform/x11/os_x11.cpp:3178
    #22 0x13ac4e1 in main platform/x11/godot_x11.cpp:56
    #23 0x7f5d835ebb96 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x21b96)

Indirect leak of 1040 byte(s) in 1 object(s) allocated from:
    #0 0x7f5d86a69f40 in realloc (/usr/lib/x86_64-linux-gnu/libasan.so.4+0xdef40)
    #1 0xcad75fa in Memory::realloc_static(void*, unsigned long, bool) core/os/memory.cpp:140
    #2 0x1ba9cf2 in CowData<wchar_t>::resize(int) core/cowdata.h:282
    #3 0x49f27ba in String::resize(int) core/ustring.h:158
    #4 0xc7d34f5 in String::operator+=(String const&) core/ustring.cpp:276
    #5 0xc7d31a4 in String::operator+(String const&) const core/ustring.cpp:252
    #6 0x4a24f0c in FileAccessUnix::_open(String const&, int) drivers/unix/file_access_unix.cpp:109
    #7 0xca1574a in FileAccess::open(String const&, int, Error*) core/os/file_access.cpp:110
    #8 0xce7bf4c in ResourceFormatSaverBinaryInstance::save(String const&, Ref<Resource> const&, unsigned int) core/io/resource_format_binary.cpp:1763
    #9 0xce85571 in ResourceFormatSaverBinary::save(String const&, Ref<Resource> const&, unsigned int) core/io/resource_format_binary.cpp:1965
    #10 0xcedbfeb in ResourceSaver::save(String const&, Ref<Resource> const&, unsigned int) core/io/resource_saver.cpp:119
    #11 0x1bd9732 in ResourceImporterOGGVorbis::import(String const&, String const&, Map<StringName, Variant, Comparator<StringName>, DefaultAllocator> const&, List<String, DefaultAllocator>*, List<String, DefaultAllocator>*, Variant*) modules/stb_vorbis/resource_importer_ogg_vorbis.cpp:106
    #12 0x4e2eec5 in EditorFileSystem::_reimport_file(String const&) editor/editor_file_system.cpp:1781
    #13 0x4e391e4 in EditorFileSystem::reimport_files(Vector<String> const&) editor/editor_file_system.cpp:1979
    #14 0x4dfd388 in EditorFileSystem::_update_scan_actions() editor/editor_file_system.cpp:590
    #15 0x4e13a71 in EditorFileSystem::_notification(int) editor/editor_file_system.cpp:1163
    #16 0x4e46b81 in EditorFileSystem::_notificationv(int, bool) (/home/rafal/Pulpit/mojgodot/bin/godot.x11.tools.64s+0x4e46b81)
    #17 0xc5f70f2 in Object::notification(int, bool) core/object.cpp:943
    #18 0x8094048 in SceneTree::_notify_group_pause(StringName const&, int) scene/main/scene_tree.cpp:973
    #19 0x8085f01 in SceneTree::idle(float) scene/main/scene_tree.cpp:519
    #20 0x1501b07 in Main::iteration() main/main.cpp:1935
    #21 0x1421219 in OS_X11::run() platform/x11/os_x11.cpp:3178
    #22 0x13ac4e1 in main platform/x11/godot_x11.cpp:56
    #23 0x7f5d835ebb96 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x21b96)

Indirect leak of 528 byte(s) in 1 object(s) allocated from:
    #0 0x7f5d86a69f40 in realloc (/usr/lib/x86_64-linux-gnu/libasan.so.4+0xdef40)
    #1 0xcad75fa in Memory::realloc_static(void*, unsigned long, bool) core/os/memory.cpp:140
    #2 0x1ba9cf2 in CowData<wchar_t>::resize(int) core/cowdata.h:282
    #3 0x49f27ba in String::resize(int) core/ustring.h:158
    #4 0xc7d34f5 in String::operator+=(String const&) core/ustring.cpp:276
    #5 0xc7d31a4 in String::operator+(String const&) const core/ustring.cpp:252
    #6 0xc7f07e3 in String::simplify_path() const core/ustring.cpp:3255
    #7 0xc65de7f in ProjectSettings::localize_path(String const&) const core/project_settings.cpp:65
    #8 0xce8553f in ResourceFormatSaverBinary::save(String const&, Ref<Resource> const&, unsigned int) core/io/resource_format_binary.cpp:1963
    #9 0xcedbfeb in ResourceSaver::save(String const&, Ref<Resource> const&, unsigned int) core/io/resource_saver.cpp:119
    #10 0x60d98da in ResourceImporterWAV::import(String const&, String const&, Map<StringName, Variant, Comparator<StringName>, DefaultAllocator> const&, List<String, DefaultAllocator>*, List<String, DefaultAllocator>*, Variant*) editor/import/resource_importer_wav.cpp:548
    #11 0x4e2eec5 in EditorFileSystem::_reimport_file(String const&) editor/editor_file_system.cpp:1781
    #12 0x4e391e4 in EditorFileSystem::reimport_files(Vector<String> const&) editor/editor_file_system.cpp:1979
    #13 0x4dfd388 in EditorFileSystem::_update_scan_actions() editor/editor_file_system.cpp:590
    #14 0x4e13a71 in EditorFileSystem::_notification(int) editor/editor_file_system.cpp:1163
    #15 0x4e46b81 in EditorFileSystem::_notificationv(int, bool) (/home/rafal/Pulpit/mojgodot/bin/godot.x11.tools.64s+0x4e46b81)
    #16 0xc5f70f2 in Object::notification(int, bool) core/object.cpp:943
    #17 0x8094048 in SceneTree::_notify_group_pause(StringName const&, int) scene/main/scene_tree.cpp:973
    #18 0x8085f01 in SceneTree::idle(float) scene/main/scene_tree.cpp:519
    #19 0x1501b07 in Main::iteration() main/main.cpp:1935
    #20 0x1421219 in OS_X11::run() platform/x11/os_x11.cpp:3178
    #21 0x13ac4e1 in main platform/x11/godot_x11.cpp:56
    #22 0x7f5d835ebb96 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x21b96)

Indirect leak of 528 byte(s) in 1 object(s) allocated from:
    #0 0x7f5d86a69f40 in realloc (/usr/lib/x86_64-linux-gnu/libasan.so.4+0xdef40)
    #1 0xcad75fa in Memory::realloc_static(void*, unsigned long, bool) core/os/memory.cpp:140
    #2 0x1ba9cf2 in CowData<wchar_t>::resize(int) core/cowdata.h:282
    #3 0x49f27ba in String::resize(int) core/ustring.h:158
    #4 0xc7d34f5 in String::operator+=(String const&) core/ustring.cpp:276
    #5 0xc7d31a4 in String::operator+(String const&) const core/ustring.cpp:252
    #6 0xc7f07e3 in String::simplify_path() const core/ustring.cpp:3255
    #7 0xc65de7f in ProjectSettings::localize_path(String const&) const core/project_settings.cpp:65
    #8 0xce8553f in ResourceFormatSaverBinary::save(String const&, Ref<Resource> const&, unsigned int) core/io/resource_format_binary.cpp:1963
    #9 0xcedbfeb in ResourceSaver::save(String const&, Ref<Resource> const&, unsigned int) core/io/resource_saver.cpp:119
    #10 0x1bd9732 in ResourceImporterOGGVorbis::import(String const&, String const&, Map<StringName, Variant, Comparator<StringName>, DefaultAllocator> const&, List<String, DefaultAllocator>*, List<String, DefaultAllocator>*, Variant*) modules/stb_vorbis/resource_importer_ogg_vorbis.cpp:106
    #11 0x4e2eec5 in EditorFileSystem::_reimport_file(String const&) editor/editor_file_system.cpp:1781
    #12 0x4e391e4 in EditorFileSystem::reimport_files(Vector<String> const&) editor/editor_file_system.cpp:1979
    #13 0x4dfd388 in EditorFileSystem::_update_scan_actions() editor/editor_file_system.cpp:590
    #14 0x4e13a71 in EditorFileSystem::_notification(int) editor/editor_file_system.cpp:1163
    #15 0x4e46b81 in EditorFileSystem::_notificationv(int, bool) (/home/rafal/Pulpit/mojgodot/bin/godot.x11.tools.64s+0x4e46b81)
    #16 0xc5f70f2 in Object::notification(int, bool) core/object.cpp:943
    #17 0x8094048 in SceneTree::_notify_group_pause(StringName const&, int) scene/main/scene_tree.cpp:973
    #18 0x8085f01 in SceneTree::idle(float) scene/main/scene_tree.cpp:519
    #19 0x1501b07 in Main::iteration() main/main.cpp:1935
    #20 0x1421219 in OS_X11::run() platform/x11/os_x11.cpp:3178
    #21 0x13ac4e1 in main platform/x11/godot_x11.cpp:56
    #22 0x7f5d835ebb96 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x21b96)

SUMMARY: AddressSanitizer: 3353 byte(s) leaked in 7 allocation(s).
```
Also Address Sanitizer shows this errors
```
thirdparty/misc/stb_vorbis.c:4491:15: runtime error: member access within misaligned address 0x7f5d47af228c for type 'struct stb_vorbis', which requires 8 byte alignment
0x7f5d47af228c: note: pointer points here
  fc 01 fc 03 44 ac 00 00  02 00 00 00 d4 51 03 00  00 10 00 00 71 02 00 00  00 00 00 00 00 00 00 00
              ^ 
thirdparty/misc/stb_vorbis.c:4499:18: runtime error: member access within misaligned address 0x7f5d47af228c for type 'struct stb_vorbis', which requires 8 byte alignment
0x7f5d47af228c: note: pointer points here
  fc 01 fc 03 44 ac 00 00  02 00 00 00 d4 51 03 00  00 10 00 00 71 02 00 00  00 00 00 00 00 00 00 00
              ^ 
thirdparty/misc/stb_vorbis.c:4513:25: runtime error: member access within misaligned address 0x7f5d47af228c for type 'struct stb_vorbis', which requires 8 byte alignment
0x7f5d47af228c: note: pointer points here
  fc 01 fc 03 44 ac 00 00  02 00 00 00 d4 51 03 00  00 10 00 00 71 02 00 00  00 00 00 00 00 00 00 00
```
**Steps to reproduce:**
Run project without .import folder 

**Minimal reproduction project:**
https://github.com/godotengine/godot-demo-projects/tree/master/2d/dodge_the_creeps

### PR DESCRIPTION
**Pull request by qarmin**
Fixes #31120 
Fixes #30658

### CODE REVIEW

File: core/io/resource_format_binary.cpp [hunk 1]
```diff
@@ -1785,7 +1785,7 @@ Error ResourceFormatSaverBinaryInstance::save(const String &p_path, const RES &p
 	f->store_32(FORMAT_VERSION);
 
 	if (f->get_error() != OK && f->get_error() != ERR_FILE_EOF) {
-		f->close();
+		memdelete(f);
```
Comment (akien-mga) [inline core/io/resource_format_binary.cpp]: `f->close();` should still be kept, it's important for the OS that we release the file handle properly.

Comment (qarmin) [reply core/io/resource_format_binary.cpp]: Done

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Comment by akien-mga** [comment]
Cherry-picked for 3.1.2.

**Pull request merged**